### ISSUE TITLE
only short outputs

### ISSUE THREAD

**Issue by fourteatoo** [rubencaro/sshex#4]
The following code results, after some output, into a timeout error:

  def test_ssh_stream() do
    {:ok, conn} = :ssh.connect('somehost', 22, [user: 'someuser'])
    stream = SSHEx.stream(conn, 'yes')
    Enum.each(stream, fn x -> IO.puts ">>> #{inspect x}" end)
  end

[...]
ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\ny\n"}

> > > {:error, "Timeout. Did not receive data for 5000ms."}
> > > :ok
> > > iex(24)> 

The code seems to hang waiting for a message that never comes.
Is there something I'm missing?

**Comment by rubencaro** [comment]
First, I guess it's a typo, but you should use `Stream.each/2` instead of `Enum.each/2`.  I'm not sure if it has anything to do with it anyway.

That timeout comes from https://github.com/rubencaro/sshex/blob/master/lib/sshex.ex#L189 . It means the process did not receive any data message from the ssh connection channel for 5secs. It may mean the connection is actually jammed, or some intermediate process has some kind of problem. If you could take a look at the processes involved, using `:observer` for example, while they are running, you may see what the problem is.

**Comment by fourteatoo** [comment]
No it is not a typo.  Stream.each just returns another stream (augmented by
its function), while Enum.each does actually realize the stream and apply
the function.
Yes, I did see where the timeout happens, but I wasn't able to figure out
why.  The problem is reproducible (ie, it happens always) at least on OSX,
and the remote host is a virtual machine.  I'd rule out connection problems.

**Comment by rubencaro** [comment]
I got it reproduced too. Will take a look at it when I find some time!

**Comment by rubencaro** [comment]
I'm playing with it now. 

This works ok: `SSHEx.stream(conn,'while sleep 0.0001; do echo hello; done;') |> Enum.each(fn(x)-> x |> inspect |> IO.puts end)`

This does not: `SSHEx.stream(conn,'while true; do echo hello; done;') |> Enum.each(fn(x)-> x |> inspect |> IO.puts end)`

So it's not about the _length_ but about the _speed_ of output production.

**Comment by rubencaro** [comment]
More mental notes:

Pieces are `SSHEx` -> `:ssh` -> `Lower level SSH` -> `command`. 

When I run `SSHEx.stream(conn,'while true; do echo hello; done;') |> Enum.each(fn(x)-> x |> inspect |> IO.puts end)` the first status is this: `SSHEx` keeps receiving data messages from `:ssh`, and `command` appears as running on remote machine. 

At some point `SSHEx` stops receiving messages from `:ssh`. Some time _after_ that, `command` stops running on the remote machine (I guess it fills it's local output buffer). It looks like some buffer is not being consumed somewhere in `:ssh` or in `Lower level SSH`.

**Comment by rubencaro** [comment]
This is the message that stops being sent https://github.com/erlang/otp/blob/82a835d94be7ee5e98d101a29999fedaf6cd75fe/lib/ssh/src/ssh_connection_handler.erl#L1789

**Comment by fourteatoo** [comment]
Nope it is not a matter of speed.
With the sleep it just takes longer, but it will eventually go in timeout all the same.

**Comment by rubencaro** [comment]
You are absolutely right. The speed counts, but it's not the limit. I managed to get the length of the complete response output received in all the messages. 

What I get is that if the _speed_ is too much, then the piece of data on each message is bigger. So I receive few long messages before it stops. When the speed is low, then I receive many small packages. 

What I also found out is that the total length of data received seems to tend to 65535 bytes when the speed is low enough and the size of the messages is small enough too.

`String.length` over the accumulated responses from:
- `while true; do echo hello; done;` :  18836
- `while sleep 0.01; do echo hello; done;`: 65532
- `while sleep 0.01; do echo; done;`: 65535

I'm asking about this on erlang mailing list. There must be something they know.
